File: Difficulty: Medium/The Celebrity Problem/the-celebrity-problem.java
//{ Driver Code Starts
//Initial Template for Java

import java.io.*;
import java.util.*; 

class GFG{
    public static void main(String args[]) throws IOException { 
        Scanner sc = new Scanner(System.in);
        int t = sc.nextInt();
        while(t>0)
        {
            int N = sc.nextInt();
            int M[][] = new int[N][N];
            for(int i=0; i<N; i++)
            {
                for(int j=0; j<N; j++)
                {
                    M[i][j] = sc.nextInt();
                }
            }
            System.out.println(new Solution().celebrity(M,N));
            t--;
        }
    } 
} 
// } Driver Code Ends


class Solution
{ 
    //Function to find if there is a celebrity in the party or not.
    int celebrity(int M[][], int n)
    {
    	int start = 0;
    	int end = n-1;
    	while(start<end){
    	    if(M[start][end]==1) start++;
    	    else if(M[end][start]==1) end--;
    	    else{
    	        start++;
    	        end--;
    	    }
    	}
    	if(start==end){
    	    for(int i=0;i<n;i++){
    	        if(i!=start && (M[i][start]!=1 || M[start][i]!=0))
    	            return -1;
    	    }
    	}
    	return start;
    	
    }
}

File: Difficulty: Medium/The Celebrity Problem/the-celebrity-problem.cpp
//{ Driver Code Starts
// Initial template for C++

#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends
// User function template for C++

class Solution {
  public:
    int celebrity(vector<vector<int> >& M) 
    { 
        int n=M.size(); 
        
        stack<int>s; 
        for(int i=0; i<n; i++){
            s.push(i); 
        }
        while(s.size()>1){
            int i=s.top();
            s.pop(); 
            int j=s.top(); 
            s.pop(); 
            if(M[i][j]){
                s.push(j); 
            }
            else{
                s.push(i); 
            }
        }
        for(int i=0; i<n; i++){
            if(i!=s.top()){
                if(M[s.top()][i] or !M[i][s.top()]){
                    return -1; 
                }
            }
        }
        
        return s.top(); 
        
    }
};

//{ Driver Code Starts.

int main() {
    int t;
    cin >> t;
    while (t--) {
        int n;
        cin >> n;
        vector<vector<int> > M(n, vector<int>(n, 0));
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                cin >> M[i][j];
            }
        }
        Solution ob;
        cout << ob.celebrity(M) << endl;
    }
}

// } Driver Code Ends